ingestion/ingest-google-docs.py

```diff
@@ -0,0 +1,134 @@
+from __future__ import print_function
+import json
+import pickle
+import os.path
+from time import sleep
+from apiclient import discovery
+from httplib2 import Http
+from oauth2client import client
+from oauth2client import file
+from oauth2client import tools
+from googleapiclient.discovery import build
+from google_auth_oauthlib.flow import InstalledAppFlow
+from google.auth.transport.requests import Request
+from git import Repo
+
+SCOPES = ['https://www.googleapis.com/auth/drive.metadata.readonly']
+SCOPES_DOC = ['https://www.googleapis.com/auth/documents.readonly']
+DISCOVERY_DOC = ('https://docs.googleapis.com/$discovery/rest?'
+                         'version=v1')
+
+ROOT_PATH = os.path.abspath(os.path.join(os.path.dirname( __file__ ), '..'))
+GIT_REPO_PATH = f'{ROOT_PATH}/.git'
+COMMIT_MESSAGE = 'Update shared repository'
+
+def git_push():
+    repo = Repo(GIT_REPO_PATH)
+    repo.index.add(["shared_directory"])
+    repo.index.commit(COMMIT_MESSAGE)
+    repo.git.push('origin', 'feature_docs_api') 
+
+def read_paragraph_element(element):
+    """Returns the text in the given ParagraphElement.
+
+        Args:
+            element: a ParagraphElement from a Google Doc.
+    """
+    text_run = element.get('textRun')
+    if not text_run:
+        return ''
+    return text_run.get('content')
+
+
+def read_strucutural_elements(elements):
+    """Recurses through a list of Structural Elements to read a document's text where text may be
+        in nested elements.
+
+        Args:
+            elements: a list of Structural Elements.
+    """
+    text = ''
+    for value in elements:
+        if 'paragraph' in value:
+            elements = value.get('paragraph').get('elements')
+            for elem in elements:
+                text += read_paragraph_element(elem)
+        elif 'table' in value:
+            # The text in table cells are in nested Structural Elements and tables may be
+            # nested.
+            table = value.get('table')
+            for row in table.get('tableRows'):
+                cells = row.get('tableCells')
+                for cell in cells:
+                    text += read_strucutural_elements(cell.get('content'))
+        elif 'tableOfContents' in value:
+            # The text in the TOC is also in a Structural Element.
+            toc = value.get('tableOfContents')
+            text += read_strucutural_elements(toc.get('content'))
+    return text

```

Comment:
Can you write a unit test for read_structural_elements ?

Reply:
Made a skeleton for it, but the actual format of the JSON that will be written to the shared directory will probably change soon (as discussed, preserve the formatting of the doc) so I'll hold off for now

Reply:
Sounds good.